### PR TITLE
Menu refactor

### PR DESCRIPTION
**Pull request by rn0**

### CODE REVIEW

File: Menu/Builder/YamlBuilder.php [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+        $this->manager = $manager;
+    }
+
+    /**
+     * @return Item
+     */
+    public function buildMenu()
+    {
+        $config = $this->yaml->parse($this->configFilePath, true, true);
+        $menuConfig = $config['menu'];
```
Comment (bartekgruszka) [inline Menu/Builder/YamlBuilder.php]: This variable is redundant

**Comment by bartekgruszka** [comment]
:+1:

File: Menu/Item/Item.php [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+    {
+        $optionsResolver = new OptionsResolver();
+        $this->setDefaultOptions($optionsResolver);
+        $this->options = $optionsResolver->resolve($options);
+    }
+
+    /**
+     * @param OptionsResolverInterface $resolver
+     */
+    public function setDefaultOptions(OptionsResolverInterface $resolver)
```
Comment (rn0) [inline Menu/Item/Item.php]: public -> private

**Comment by rn0** [comment]
This PR solves https://github.com/fsi-open/admin-translatable-bundle/issues/11

File: Event/MenuBuilderEvent.php [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+ */
+
+namespace FSi\Bundle\AdminBundle\Event;
+
+use FSi\Bundle\AdminBundle\Menu\Item\Item;
+use Symfony\Component\EventDispatcher\Event;
+
+class MenuBuilderEvent extends Event
+{
+    const TOOLS = 'fsi_admin.menu.tools';
```
Comment (chives) [inline Event/MenuBuilderEvent.php]: I would extract this constant to separate ``MenuEvents`` class and rename this class ``MenuEvent``

File: EventListener/MenuToolsListener.php [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+ */
+
+namespace FSi\Bundle\AdminBundle\EventListener;
+
+use FSi\Bundle\AdminBundle\Event\MenuBuilderEvent;
+use FSi\Bundle\AdminBundle\Menu\Item\RoutableItem;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\Translation\TranslatorInterface;
+
+class MenuToolsListener
```
Comment (chives) [inline EventListener/MenuToolsListener.php]: ``LocaleMenuListener`` would be better

File: Menu/Builder/EventedBuilder.php [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+ * file that was distributed with this source code.
+ */
+
+namespace FSi\Bundle\AdminBundle\Menu\Builder;
+
+use FSi\Bundle\AdminBundle\Event\MenuBuilderEvent;
+use FSi\Bundle\AdminBundle\Menu\Item\Item;
+use Symfony\Component\EventDispatcher\EventDispatcherInterface;
+
+class EventedBuilder implements Builder
```
Comment (chives) [inline Menu/Builder/EventedBuilder.php]: As long as it builds only the tools menu it should be named something like ``ToolsMenuBuilder``

**Comment by boekkooi** [comment]
@rn0 It looks like  admin-bundle/Resources/views/Admin/main_menu_theme.html.twig needs to be updated to use the `matcher` see https://github.com/KnpLabs/KnpMenu/blob/master/src/Knp/Menu/Resources/views/knp_menu.html.twig#L55

Also a custom Matcher\Voter is probably needed due to the fact that `fsi_admin_crud_list` are currently not matched when you on the list because the route itself is `fsi_admin_list` I would use `Knp\Menu\Matcher\Voter\RouteVoter` as a base here.

Also I think it would be better to prefix the menu names with `fsi_admin_` since else it may conflict when integrating the bundle into existing projects (aka `<tag name="knp_menu.menu" alias="main"/>` would become `<tag name="knp_menu.menu" alias="fsi_admin_main"/>` etc.).

**Comment by chives** [comment]
:+1:

File: EventListener/MainMenuListener.php [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+    private $manager;
+
+    /**
+     * @param ManagerInterface $manager
+     * @param string $configFilePath
+     */
+    public function __construct(ManagerInterface $manager, $configFilePath)
+    {
+        $this->configFilePath = $configFilePath;
+        $this->yaml = new Yaml();
```
Comment (bartekgruszka) [inline EventListener/MainMenuListener.php]: Dependency Injection?

File: Menu/Builder/Exception/InvalidYamlStructure.php [hunk 7]
```diff
@@ -0,0 +1,8 @@
+<?php
+
+namespace FSi\Bundle\AdminBundle\Menu\Builder\Exception;
+
+class InvalidYamlStructure extends \Exception
```
Comment (bartekgruszka) [inline Menu/Builder/Exception/InvalidYamlStructure.php]: `InvalidYamlStructureException`

**Comment by bartekgruszka** [comment]
:+1:

**Pull request merged**